### PR TITLE
[Configurator] Transpile CSS on Hash Change, and other small fixes / offline requests

### PR DESCRIPTION
**Pull request by torch2424**
* Introduce the CSSTranspiler Class
* Restructure code for better readability
* Add postcss and postcss-custom-properties as dependency
* Restructure code for asynchronous initialization
* Update CSS on Hash Change

# Example Gif

configuratorhash

### CODE REVIEW

File: www/configurator/src/app/iframe-manager/iframe-manager.js [hunk 1]
```diff
@@ -23,39 +23,37 @@
...
-    this.style = this.iframe.contentDocument.createElement('style');
-    this.iframe.contentDocument.head.appendChild(this.style);
+  initialize() {
+    return new Promise(resolve => {
+      // Set the iframe src
+      this.iframe.src = this.templateSrc;
+      // Create our style object we will be using to overide styles in the iframe
+      this.iframe.addEventListener('load', () => {
+        this.styleElement = this.iframe.contentDocument.createElement('style');
+        this.iframe.contentDocument.head.appendChild(this.styleElement);
```
Comment (camelburrito) [inline www/configurator/src/app/iframe-manager/iframe-manager.js]: this has to replace amp-custom style tag. Where is this element instantiated?

File: www/configurator/src/app/iframe-manager/iframe-manager.js [hunk 2]
```diff
@@ -23,39 +23,37 @@
...
-  }
-
-  setSrc_(path, iframeLoadedCallback) {
-    this.iframe.src = path;
-    this.iframe.addEventListener('load', () => {
-      if (iframeLoadedCallback) {
-        iframeLoadedCallback();
-      }
-    });
+    this.styleElement.textContent = style;
```
Comment (camelburrito) [inline www/configurator/src/app/iframe-manager/iframe-manager.js]: look at comment above

File: www/configurator/src/index.js [hunk 3]
```diff
@@ -33,36 +33,63 @@ if (process.env.NODE_ENV === 'production') {
...
+  params.hash = Object.assign({}, params.hash, queryString.parse(location.hash.substring(1)));
+  return params;
+}
+
+// Define our iframe manager, an our css transplier
+let iframeManager = false;
+let cssTranspiler = false;
+// Get our url params on page load
+let params = getUrlParams();
+console.log('Initial Params', params);
```
Comment (camelburrito) [inline www/configurator/src/index.js]: remove?

File: www/configurator/src/index.js [hunk 4]
```diff
@@ -33,36 +33,63 @@ if (process.env.NODE_ENV === 'production') {
   cssPath = 'test-dist/uncompiled-css/templates/';
 }
 
+// Function to get params on the page
+function getUrlParams() {
```
Comment (camelburrito) [inline www/configurator/src/index.js]: split it into 2 methods, one to get the template and other to get the URL hash.

Comment (camelburrito) [inline www/configurator/src/index.js]: can you js doc all the functions - all this code is getting hard to read.

File: www/configurator/src/index.js [hunk 5]
```diff
@@ -33,36 +33,63 @@ if (process.env.NODE_ENV === 'production') {
...
+  const templateParam = queryString.parse(location.search.substring(1)).template;
+  if (templateParam) {
+    params.search.template = templateParam;
+  }
+  params.hash = Object.assign({}, params.hash, queryString.parse(location.hash.substring(1)));
+  return params;
+}
+
+// Define our iframe manager, an our css transplier
+let iframeManager = false;
```
Comment (camelburrito) [inline www/configurator/src/index.js]: I guess they can be declared where needed , it is confusing when you declare them here and instantiate somewhere else.

File: www/configurator/src/index.js [hunk 6]
```diff
@@ -33,36 +33,63 @@ if (process.env.NODE_ENV === 'production') {
...
-cssRequests.push(
+configuratorInit.push(
   fetch(`${templateCssPath}.css`).then(response => {
     return response.text();
   })
 );
 
-Promise.all(cssRequests).then(responses => {
+// Create the configurator
+const templateSrc = `${templatesPath}${params.search.template}/${params.search.template}.amp.html#amp=1`;
```
Comment (camelburrito) [inline www/configurator/src/index.js]: templatesPath- Where is this being set?

File: www/configurator/src/index.js [hunk 7]
```diff
@@ -33,36 +33,63 @@ if (process.env.NODE_ENV === 'production') {
...
+
+// Define our iframe manager, an our css transplier
+let iframeManager = false;
+let cssTranspiler = false;
+// Get our url params on page load
+let params = getUrlParams();
+console.log('Initial Params', params);
+
 // Define our hash change handler
 function handleHashChange_() {
```
Comment (camelburrito) [inline www/configurator/src/index.js]: move this function below your promise.

File: www/configurator/src/app/iframe-manager/iframe-manager.js (hunk 1 above)
Comment (torch2424) [inline www/configurator/src/app/iframe-manager/iframe-manager.js]: Was planning on doing this in another PR. But I can do it here.

Done

File: www/configurator/src/app/iframe-manager/iframe-manager.js (hunk 2 above)
Comment (torch2424) [inline www/configurator/src/app/iframe-manager/iframe-manager.js]: Done

File: www/configurator/src/index.js (hunk 4 above)
Comment (torch2424) [inline www/configurator/src/index.js]: Was planning on doing this in another PR too keep the CL size low, but I can definitely do it here :)

File: www/configurator/src/index.js (hunk 6 above)
Comment (torch2424) [inline www/configurator/src/index.js]: It's being set underneath the imports. `templatesPath` is conditionally set based on the current environment variable that webpack exposes when it builds

File: www/configurator/src/index.js (hunk 3 above)
Comment (torch2424) [inline www/configurator/src/index.js]: Done

File: www/configurator/src/index.js (hunk 7 above)
Comment (torch2424) [inline www/configurator/src/index.js]: Done

File: www/configurator/src/index.js (hunk 5 above)
Comment (torch2424) [inline www/configurator/src/index.js]: Done

File: www/configurator/src/index.js (hunk 4 above)
Comment (torch2424) [inline www/configurator/src/index.js]: Done

Comment (torch2424) [inline www/configurator/src/index.js]: Done

File: www/configurator/src/app/iframe-manager/iframe-manager.spec.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const assert = require('assert');
+
+describe('karma testing', () => {
```
Comment (camelburrito) [inline www/configurator/src/app/iframe-manager/iframe-manager.spec.js]: you will need to start adding tests soon :)

**Pull request merged**